packages/replay-next/src/hooks/useNotification.ts

```diff
@@ -0,0 +1,60 @@
+import { useCallback, useEffect, useState } from "react";
+
+export type Permission = NotificationPermission | PermissionState;
+export type RequestPermission = () => Promise<boolean>;
+
+export function useNotification(): {
+  permission: Permission;
+  requested: boolean;
+  requestPermission: RequestPermission;
+  supported: boolean;
+} {
+  const [permission, setPermission] = useState<Permission>(Notification.permission);
+
+  useEffect(() => {
+    let permissionStatus: PermissionStatus;
+
+    const onChange = () => {
+      if (permissionStatus) {
+        setPermission(permissionStatus.state);
+      }
+    };
+
+    (async () => {
+      permissionStatus = await navigator.permissions.query({ name: "notifications" });
+      permissionStatus.addEventListener("change", onChange);
+    })();
+
+    return () => {
+      if (permissionStatus) {
+        permissionStatus.removeEventListener("change", onChange);
+      }
+    };
```

Comment:
There is a small race condition here. It's possible for the cleanup of this effect to be called before the query resolves. In such a scenario, we'd still call `addEventListener` and never clean it up.

I'd introduce a `disposed` variable, set it in the cleanup and check in that `async` function to handle this scenario

Reply:
Yeah, this was a little half-assed. I should have been using `useSyncExternalStore` for this. Will fix.